### DISCUSSION TITLE
linux平台下 pag文件 输出的 图片 不流畅

### DISCUSSION THREAD

**Discussion by seu070507** [Tencent/libpag#2390] [General] [answered]
L
[Uploading pag_issue.zip…]()
inux 平台下，稍微修改了demo的code，pag文件输出的图片，或者合成的视频卡顿不流畅， 怀疑pag解析过程中有丢帧问题，
但也可能是我门使用的方式问题，还请解答一下！

主要函数： 
  MakeOffscreen
  pagPlayer->setProgress
  pagPlayer->flush()
  pagSurface->readPixels

代码片段：
`int processPagFile(std::shared_ptr<pag::PAGFile> pagFile, const char* outfile, int imageWidth, int offset, int64_t startTime)
{
  if (pagFile == nullptr) {
    printf("libpag: processPagFile pagFile is nullptr!!!\n");
    return -1;
  }

  printf("libpag: processPagFile start.\n");

  int actualWidth = pagFile->width();

  if (imageWidth > 0)
    actualWidth = imageWidth;
  int actualHeight = actualWidth * pagFile->height() / pagFile->width();
  auto pagSurface = pag::PAGSurface::MakeOffscreen(actualWidth, actualHeight);
  if (pagSurface == nullptr) {
    printf("libpag: processPagFile pagSurface is nullptr!!!\n");
    return -1;
  }
  auto pagPlayer = new pag::PAGPlayer();
  pagPlayer->setSurface(pagSurface);
  pagPlayer->setComposition(pagFile);

  auto totalFrames = TimeToFrame(pagFile->duration(), pagFile->frameRate());
  auto currentFrame = 0;

  int bytesLength = actualWidth * actualHeight * 4;
  int writeIndex = offset;

  // check whether dir is exist
  std::string outdir = "";
  outdir.append(outfile);

  if (access(outdir.c_str(), 0) == -1) {
    int flag = mkdir(outdir.c_str(), S_IRWXU);
  }

  auto audioData = pagFile->audioBytes();
  AudioWrite(audioData, outdir.c_str());

  while (currentFrame <= totalFrames) {
    pagPlayer->setProgress(currentFrame * 1.0 / totalFrames);
    auto status = pagPlayer->flush();
    printf("---currentFrame:%d, flushStatus:%d \n", currentFrame, status);
    auto data = new uint8_t[bytesLength];
    pagSurface->readPixels(pag::ColorType::BGRA_8888, pag::AlphaType::Premultiplied, data,
                           actualWidth * 4);

    writeIndex++;
    std::string imageName = std::to_string(writeIndex);
    BmpWrite(data, actualWidth, actualHeight, outdir.c_str(), imageName.c_str());

    delete[] data;

    currentFrame++;
  }

  delete pagPlayer;

  printf("libpag: processPagFile timeCost:%fs \n", (GetTimer() - startTime) * 1.000f / 1000000);
  printf("libpag: processPagFile writeIndex:%d \n", writeIndex);

  return writeIndex;
}`

**Comment by domchen** [comment, marked as answer]
这是个while循环，根本没有实时渲染，不存在丢帧的概念。你注意一下素材的帧率，和浮点误差。应该是计算的progress有偏移导致定位错了帧。用PAGPlayer.nextFrame()会更加准确。
